### ISSUE TITLE
I can model vertical pools

### ISSUE THREAD

**Issue by lisps** [bpmn-io/bpmn-js#507]
__Describe the issue__

Vertical pools is a feature request we see once in a while. It is also part of [the MIWG test suite, case `A.4.1`](https://github.com/bpmn-io/bpmn-miwg-test-suite/blob/master/Reference/A.4.1.bpmn) specifically.

We do not account for pool direction when rendering it (https://github.com/bpmn-io/bpmn-js/issues/57) and do not allow modeling it, too.

__Tasks__

* [ ] Allow vertical pools to be configured (how does user decide whether to model vertically or horizontal)?
* [ ] Make features vertical pool aware:
    * [ ] Auto-Place
    * [ ] _?_

---

Referenced [in forum](https://forum.bpmn.io/t/support-for-vertical-modeling-pool/826/8).

**Comment by gimbel** [comment]
At the moment vertical pools are not possible. 

Are you interested in providing a pull request for that?

**Comment by lisps** [comment]
my rescourses are limited, would it be much work?

**Comment by gimbel** [comment]
Well, I guess it will require some effort, also in order to get into the project

**Comment by matthiasg** [comment]
We might be interested in any looking at doing this. Can anybody from core same give  some pointers?

**Comment by nikku** [comment]
Sorry for not getting back to you earlier. Did you already have the chance to get familiar with diagram-js/bpmn-js?

**Comment by matthiasg** [comment]
No. We have a customer requiring vertical lanes, but we are currently trying to gauge how much of a hard requirement it really is. I have extensive experience doing similar graphs in our own apps though.

**Comment by lisps** [comment]
@matthiasg has you worked on this?

**Comment by nikku** [comment]
This is not on our roadmap. If someone wants to contribute ---  we accept PRs.

**Comment by matthiasg** [comment]
@lisps no, customer prioritized poster print support instead. 

btw: I did not see your comment until now. sorry for that

**Comment by nikku** [comment]
We'll not support vertical collaboration, unless we receive a contribution for it.

Related we may implement proper rendering of vertical pools (https://github.com/bpmn-io/bpmn-js/issues/57) some time maybe.

**Comment by sombrek** [comment]
@nikku I'd like to contribute to this feature.

I think I can implement resizing via drag&drop and using the space tool.
Is that an acceptable scope for an increment?

**Comment by nikku** [comment]
@sombrek As an initial step I'd suggest you to see what is currently broken, and summarize it here. We can then go ahead and give you pointers where to fix things.

**Comment by sombrek** [comment]
@nikku Here's a summary of what I've got so far.

- vertical pools and lanes still use the minimum dimensions of horizontal pools
- resizing vertical elements using resize handles causes overlapping elements

<details>
  <summary>Resize Issue Demonstration</summary>

![pool-issues-resize](https://github.com/bpmn-io/bpmn-js/assets/150448993/165951ba-e786-4a5e-9bb6-a88d8f9c996c)

</details>

I also noticed some existing quirks for _horizontal_ pools and I would like your input on this.
- Black-box pools can be resized in any direction using handles, but the space tool only allows resizing them horizontally. That could be intentional, but I don't know.
- The resize handles use minimum dimensions of 300 x 60 for pools and lanes, while the space tool has a minimum of 300 x 150 for pools. Should vertical elements use the same logic, just switch width and height? Or should pools/lanes have the same dimensions for both tools (if so which should be used)?

I have identified `lib/features/modeling/behavior/ResizeBehavior.js`, `lib/features/modeling/behavior/SpaceToolBehavior.js`, and maybe `lib/features/modeling/util/LaneUtil.js` as startig points for an implementation.

Other areas of work:
- Adding/Splitting should create vertical lanes within vertical lanes
- Placing new tasks/events on the vertical axis when modeling within a vertical pool (is that meant with "Auto-Place" in the description?).

I think full support also needs new icons...
- Icons for adding/splitting lanes vertically
- And, depending on whether you want to expose those elements/features on the UI
  - Vertical pool and black-box pool
  - Icons to switch from horizontal to vertical pool and vice versa

... and new translatable texts. In both cases I don't know the extra work that might be required.

**Comment by philippfromme** [comment]
You've done a good job identifying most aspects of making vertical pools work. As you can see, there are many. I'd suggest starting with the smallest possible increment which I think would be:

* we only deal with existing vertical pools and lanes and don't want to create new ones for now
* resizing of the pools and lanes works as expected
  * minimum dimensions are set appropriately
  * space tool resizes vertical pools and lanes appropriately

For that, I'd suggest creating a separate issue to track this so this issue can serve as the epic.

What do you think?

**Comment by sombrek** [comment]
There was a lot of progress regarding this feature.

All tasks in "vertical pools and lanes can be updated" should work now, except for "pools can be switched between horizontal and vertical". There haven't been any UI updates to reflect this feature.

"Vertical modeling is enabled in vertical pools", i.e. auto-place does work. However, it is not perfect yet, due to issue #2127.
Moreover, some minor features were adjusted to consider vertical pools (e.g. hit boxes).

I'm giving this update, because some time may pass until my next contribution. Thanks for all your kind and helpful support.

**Comment by barmac** [comment]
Thank you for your amazing work on this epic. I updated the top comment according to the progress delivered.

**Comment by nikku** [comment]
@sombrek Some amazing contributions, indeed. Will make some users happy :cake: :1st_place_medal:

**Comment by rip222** [comment]
This is amazing! Looking forward to creating pools and lanes. Been waiting for it for years

**Comment by sombrek** [comment]
@rip222 At the moment there will be no progress from my side.

If you simply want to add a vertical pool to your diagram, I recommend the [steps I described in the forum](https://forum.bpmn.io/t/vertical-pool-in-camunda-modeler/10376/8). Once your pool is vertical, most existing features will work as expected.

If you'd like to try implementing more features, here are some starting points.

I'm not 100 % sure, but to create vertical pools via code, you would have to extend the `ElementFactory` so that you can set the `isHorizontal` attribute accordingly. https://github.com/bpmn-io/bpmn-js/blob/abd8c6c1c2435576b5b8b24275405ba32cc51927/lib/features/modeling/ElementFactory.js#L337-L351

To create pools using the modeler, you might have (in addition to the above) create a plugin that provides this feature as an extension of the [PaletteProvider](https://github.com/bpmn-io/bpmn-js/blob/develop/lib/features/palette/PaletteProvider.js). I say plugin, because the maintainers may not want this feature available by default.

Regarding updating: You'd have to write new code that translates horizontal pools and lanes into vertical ones (and vice versa), but not their content (at least not as part of the first iteration). This could be made available in the [ContextPadProvider](https://github.com/bpmn-io/bpmn-js/blob/develop/lib/features/context-pad/ContextPadProvider.js) (possibly as a plugin as well).

If you want to pick up any of this, simply [create an issue](https://github.com/bpmn-io/bpmn-js/issues/new/choose) to get in contact with the maintainers. You can see examples of how I did it in some of the linked issues above my previous post.

**Comment by nikku** [comment]
@sombrek Thanks for the guidance :heart:

@rip222 As a maintainer I fully subscribe to https://github.com/bpmn-io/bpmn-js/issues/507#issuecomment-2088490846. 

Specifically, whether you want to model vertically _or_ horizontally may be a decision that you take upfront (editor setting, or that is [inferred from the context](https://github.com/bpmn-io/design-principles?tab=readme-ov-file#know-the-context); we do not aim to provide you the option to switch back and forth during editing in the out-of-the-box experience.
